File: leetcode/weekly-contest/20240316/100262-求出加密整数的和.cpp
#include <vector>

class Solution
{
public:
  int sumOfEncryptedInt(std::vector<int> &nums)
  {
    int sum = 0;
    for (int num : nums) {
      int len = 0;
      int max  = -1;
      // find max and len
      while (num > 0) {
        max   = std::max(num % 10, max);
        num /= 10;
        ++len;
      }

      // encode
      num = 0;
      while (len > 0) {
        num = num * 10 + max;
        --len;
      }
      sum += num;
    }
    return sum;
  }
};

File: leetcode/weekly-contest/20240316/100262-求出加密整数的和.py
from typing import List

class Solution:
    def sumOfEncryptedInt(self, nums: List[int]) -> int:
        # def f(x):
        #     s = str(x)
        #     return int(max(s) * len(s))

        def f(x):
            cnt, v = 0, 0
            while x:
                x, tmp = divmod(x, 10)
                cnt += 1
                v = max(v, tmp)
            # 最后相当于乘以 11...1 ，也就是 99...9 除以 9
            return v * (10 ** cnt - 1) // 9

        return sum(f(x) for x in nums)
